src/forest.rs

```diff
@@ -243,10 +241,324 @@ impl Forest {
+        let c2_id = add_default_leaf(&mut forest);
+        forest.add_child(parent_id, c1_id);
+        forest.add_child(parent_id, c2_id);
+
+        let moved_id = forest.swap_remove(c1_id);
+        let new_c2_id = c1_id.clone();
+
+        assert_eq!(moved_id, Some(c2_id));
+        assert_eq!(forest.children[parent_id].len(), 1);
+        assert_eq!(forest.children[parent_id][0], new_c2_id);
+        assert_eq!(forest.parents[new_c2_id].len(), 1);
+        assert_eq!(forest.parents[new_c2_id][0], parent_id);
+        assert_forest_size(&forest, 2);
+    }
+
+    #[test]
+    fn swap_remove_last_child_nested() {
+        let mut forest = Forest::with_capacity(3);
+        let parent_id = add_default_leaf(&mut forest);
+        let c1_id = add_default_leaf(&mut forest);
+        let c2_id = add_default_leaf(&mut forest);
+        forest.add_child(parent_id, c1_id);
+        forest.add_child(parent_id, c2_id);
+
+        let moved_id = forest.swap_remove(c2_id);
+
+        assert_eq!(moved_id, None);
+        assert_eq!(forest.children[parent_id].len(), 1);
+        assert_eq!(forest.parents[c1_id].len(), 1);
+        assert_forest_size(&forest, 2);
+    }
+
+    #[test]
+    fn swap_remove_disjoint() {
+        let mut forest = Forest::with_capacity(2);
+        let n1_id = add_default_leaf(&mut forest);
+        let n2_id = add_default_leaf(&mut forest);
+
+        let moved_id = forest.swap_remove(n1_id);
+
+        assert_eq!(moved_id, Some(n2_id));
+        assert_forest_size(&forest, 1);
+    }
+
+    #[test]
+    fn remove_child() {
+        let mut forest = Forest::with_capacity(3);
+        let parent_id = add_default_leaf(&mut forest);
+        let c1_id = add_default_leaf(&mut forest);
+        let c2_id = add_default_leaf(&mut forest);
+        forest.add_child(parent_id, c1_id);
+        forest.add_child(parent_id, c2_id);
+
+        let removed_id = forest.remove_child(parent_id, c1_id);
+        let parent = &forest.nodes[parent_id];
+
+        assert!(parent.is_dirty);
+        assert_eq!(forest.children[parent_id].len(), 1);
+        assert_eq!(forest.parents[c1_id].len(), 0);
+        assert_eq!(forest.parents[c2_id].len(), 1);
+        assert_eq!(removed_id, c1_id);
+    }
+
+    #[test]
+    fn remove_child_at_index() {
+        let mut forest = Forest::with_capacity(3);
+        let parent_id = add_default_leaf(&mut forest);
+        let c1_id = add_default_leaf(&mut forest);
+        let c2_id = add_default_leaf(&mut forest);
+        forest.add_child(parent_id, c1_id);
+        forest.add_child(parent_id, c2_id);
+
+        let removed_id = forest.remove_child_at_index(parent_id, 0);
+        let parent = &forest.nodes[parent_id];
+
+        assert!(parent.is_dirty);
+        assert_eq!(forest.children[parent_id].len(), 1);
+        assert_eq!(forest.parents[c1_id].len(), 0);
+        assert_eq!(forest.parents[c2_id].len(), 1);
+        assert_eq!(removed_id, c1_id);
+    }
+
+    #[test]
+    fn mark_dirty() {
```

Comment:
This test verifies that dirty-propagation works as expected. (Marking parent as dirty, should mark all children as dirty). I'm not sure we need another test just for "setting node to dirty, should have the node report as dirty" but I do think we should either rename the test or add a comment conveying that tests dirty-propagation

Reply:
I've named it `mark_dirty_propagates_to_parents`, should be clearer now.